### PR TITLE
fix: Blocked time release metrics are incorrect #279

### PR DESCRIPTION
**Pull request by YuliaMittova**
Bugfix for issue #279

### CODE REVIEW

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 1]
```diff
@@ -144,12 +144,19 @@ private ReleaseMetricsResponse.BlockedTime buildBlockedTime(List<Feature> featur
...
+        }
+
+        final int totalBlockedDays = blockedFeatures.stream()
+                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(f.getUpdatedAt(), Instant.now())))
+                .sum();
+        final int totalDaysSinceCreated = features.stream()
+                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(f.getCreatedAt(), Instant.now())))
+                .sum();
 
-        int totalBlockedDays = blockedFeatures.size() * 5; // Simplified calculation
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java:148]: Lol, I just love how AI produced "Simplified calculations" 😂

**Review by innokenty — APPROVED** [review]

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 2]
```diff
@@ -144,12 +144,19 @@ private ReleaseMetricsResponse.BlockedTime buildBlockedTime(List<Feature> featur
...
                 .toList();
+        if (blockedFeatures.isEmpty()) {
+            return new ReleaseMetricsResponse.BlockedTime(0, 0.0, 0.0, Collections.emptyMap());
+        }
+
+        final int totalBlockedDays = blockedFeatures.stream()
+                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(f.getUpdatedAt(), Instant.now())))
+                .sum();
+        final int totalDaysSinceCreated = features.stream()
+                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(f.getCreatedAt(), Instant.now())))
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java]: According to comment from @ElizavetaP , I think we just need to change this line here:

```suggestion
                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(f.getCreatedAt(), f.getPlanningStatus() == FeaturePlanningStatus.DONE ? f.getUpdatedAt() : Instant.now())))
```

**Review by innokenty — CHANGES_REQUESTED** [review]
Please address the discussion on the issue:
- adjust calculation logic (suggestion provided)
- adjust tests accordingly
- replace "in between" assertions with proper value

**Comment by YuliaMittova** [comment]
@ElizavetaP @innokenty added changes. Please check.
What was done:
1. Added fix to the logic as suggested
2. Removed isBetween from tests and replaced to equals
3. Also inserted values in the test with a gap so it will be possible to measure things without ranges

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 3]
```diff
@@ -144,12 +144,21 @@ private ReleaseMetricsResponse.BlockedTime buildBlockedTime(List<Feature> featur
...
                 .toList();
+        if (blockedFeatures.isEmpty()) {
+            return new ReleaseMetricsResponse.BlockedTime(0, 0.0, 0.0, Collections.emptyMap());
+        }
 
-        int totalBlockedDays = blockedFeatures.size() * 5; // Simplified calculation
-        double averageBlockedDuration = blockedFeatures.isEmpty() ? 0.0 : 5.0; // Simplified
-        double percentageOfTime = features.isEmpty()
-                ? 0.0
-                : round((double) totalBlockedDays / (features.size() * 30) * 100, 1); // Simplified
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java:152]: rounding was lost, no tests are checking that
in fact, one test that would catch this uses `within(0.1)` to avoid failing – this is wrong.

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 4]
```diff
@@ -144,12 +144,21 @@ private ReleaseMetricsResponse.BlockedTime buildBlockedTime(List<Feature> featur
...
+        final int totalBlockedDays = blockedFeatures.stream()
+                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(f.getUpdatedAt(), Instant.now())))
+                .sum();
+        final int totalDaysSinceCreated = features.stream()
+                .mapToInt(f -> (int) (ChronoUnit.DAYS.between(
+                        f.getCreatedAt(),
+                        f.getPlanningStatus() == FeaturePlanningStatus.DONE ? f.getUpdatedAt() : Instant.now())))
+                .sum();
+
+        double averageBlockedDuration = (double) totalBlockedDays / blockedFeatures.size();
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java]: no tests make assertion on this field, and also rounding should be applied here too.
bug description doesn't mention this field currently, but it should – because it's also wrong.
I updated the bug description accordingly

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 5]
```diff
@@ -144,12 +144,21 @@ private ReleaseMetricsResponse.BlockedTime buildBlockedTime(List<Feature> featur
...
+        double percentageOfTime = (double) totalBlockedDays / totalDaysSinceCreated * 100.0;
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java]: possible division by zero when blocked features were all created today, no test for this case either

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 6]
```diff
@@ -144,12 +144,21 @@ private ReleaseMetricsResponse.BlockedTime buildBlockedTime(List<Feature> featur
         List<Feature> blockedFeatures = features.stream()
                 .filter(f -> f.getPlanningStatus() == FeaturePlanningStatus.BLOCKED)
                 .toList();
+        if (blockedFeatures.isEmpty()) {
+            return new ReleaseMetricsResponse.BlockedTime(0, 0.0, 0.0, Collections.emptyMap());
+        }
```
Comment (innokenty) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java:149]: no test for this case either

**Review by innokenty — COMMENTED** [review]
My friend Claude and I found a bunch of issues here (see comments below) – and **already fixed them!** I also adjusted the bug description to mention `averageBlockedDuration` as well.

Dataset looks good to me now, @ElizavetaP please review this updated version instead – maybe you'll find something.

Summary of changes I made:

 - Round `percentageOfTime` to 1 decimal place, matching other percentages in the class. Removed `within(0.1)` tolerance
  from the test assertion.
  - Round `averageBlockedDuration` to 1 decimal place. Added new release (IDEA-2028) with 3 blocked features producing a
  non-integer average (5.3) to test this.
  - Guard against division by zero when all features were created today (`totalDaysSinceCreated = 0`). Added release
  (IDEA-2029) with a same-day blocked feature to cover this edge case.
  - Added test for empty release (IDEA-2030, no features) exercising the early-return path.

**Review by innokenty — APPROVED** [review]

**Review by ElizavetaP — APPROVED** [review]
